lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -201,30 +247,40 @@ static ParseResult
 parseComponentSignature(OpAsmParser &parser, OperationState &result,
                         SmallVectorImpl<OpAsmParser::OperandType> &ports,
                         SmallVectorImpl<Type> &portTypes) {
-  if (parsePortDefList(parser, result, ports, portTypes))
-    return failure();
+  SmallVector<OpAsmParser::OperandType> inPorts, outPorts;
+  SmallVector<Type> inPortTypes, outPortTypes;
 
-  // Record the number of input ports.
-  size_t numInPorts = ports.size();
+  if (parsePortDefList(parser, result, inPorts, inPortTypes))
+    return failure();
 
-  if (parser.parseArrow() || parsePortDefList(parser, result, ports, portTypes))
+  if (parser.parseArrow() ||
+      parsePortDefList(parser, result, outPorts, outPortTypes))
     return failure();
 
   auto *context = parser.getBuilder().getContext();
   // Add attribute for port names; these are currently
   // just inferred from the SSA names of the component.
-  SmallVector<Attribute> portNames(ports.size());
-  llvm::transform(ports, portNames.begin(), [&](auto port) -> StringAttr {
+  SmallVector<Attribute> portNames;
+  auto getPortName = [context](const auto &port) -> StringAttr {
     StringRef name = port.name;
     if (name.startswith("%"))
       name = name.drop_front();
     return StringAttr::get(context, name);
-  });
+  };
+  llvm::transform(inPorts, std::back_inserter(portNames), getPortName);
+  llvm::transform(outPorts, std::back_inserter(portNames), getPortName);
+
   result.addAttribute("portNames", ArrayAttr::get(context, portNames));
+  result.addAttribute(
+      direction::attrKey,
+      direction::packAttribute(
+          direction::genInOutDirections(inPorts.size(), outPorts.size()),
+          context));
 
-  // Record the number of input ports.
-  result.addAttribute("numInPorts",
-                      parser.getBuilder().getI64IntegerAttr(numInPorts));
+  llvm::append_range(ports, inPorts);
```

Comment:
Nit (feel free to ignore): I don't have a computer available to try myself, but it *may* be possible to do something like `ports.append({inPorts, outPorts})`?

Reply:
Unfortunately does not seem to be the case. I was hoping for functionality like that (alongside the `concat` function i added) would be present in STLExtras.h